Add pytorch-lightning decorator to nano

zhentaocc opened the pull request:
Resolved 1st task in  #3171.

- [x] Add a base class to convert pytorch model to pytorch-lightning.
- [x] Add `Trainer.compile` for user to use.
- [x] Add a unit tests.

TheaperDeng commented on python/nano/test/test_models_lightning_support.py:
```diff
@@ -0,0 +1,57 @@
+    return getattr(torch.optim, config.get("optim", "Adam"))(model.parameters(), lr=config.get("lr", 0.001))
+
+
+@lightning_support.lightning_module(loss_creator, optimizer_creator, config)
```
**It seems to me that the `trainer.compile` has a better user experience** since you can not create a opt instance before the model instance is built, which leads to some config dict. 

Still both of the methods involve a abnormal usage (i.e. multiple-decorator and `trainer.compile`). We need to give thorough and detailed user guide and in-code-warning.

jason-dai replied:
> **It seems to me that the `trainer.compile` has a better user experience** since you can not create a opt instance before the model instance is built, which leads to some config dict in the decorator case.
> 
> Still the method involve a abnormal usage (i.e. `trainer.compile`). We need to give thorough and detailed user guide and in-code-warning.
> 
> btw, we don't need to implement `onnx=True` parameter in this PR since I have not merged the onnx PR. I will do it later after this PR has been merged.

What do you mean by abnormal usage? I think the point is
1) If the user simply has a PyTorch model, he or she can directly use it in `nano.pytorch.trainer` methods (if the default behavior of `fit`, `test`, `predict`, etc., works for him or her)

2) If the user needs more complex behavior (e.g., onnxruntime support), he or she needs to explictly convert it to pl_module, and we can provide an API based on either decorator or compile method.

TheaperDeng replied:
> > **It seems to me that the `trainer.compile` has a better user experience** since you can not create a opt instance before the model instance is built, which leads to some config dict in the decorator case.
> > Still the method involve a abnormal usage (i.e. `trainer.compile`). We need to give thorough and detailed user guide and in-code-warning.
> > btw, we don't need to implement `onnx=True` parameter in this PR since I have not merged the onnx PR. I will do it later after this PR has been merged.
> 
> What do you mean by abnormal usage? I think the point is
> 
> 1. If the user simply has a PyTorch model, he or she can directly use it in `nano.pytorch.trainer` methods (if the default behavior of `fit`, `test`, `predict`, etc., works for him or her)
> 2. If the user needs more complex behavior (e.g., onnxruntime support), he or she needs to explictly convert it to pl_module, and we can provide an API based on either decorator or compile method.

I think we can just ask all users with a nn.module to do this:
```python
model = Net()
model_pl = trainer.compile(model, loss, opt, onnx=bool)
# then use model_pl (pl.lightningmodule) to do anything else
```
abnormal usage means that `trainer.compile` is not a easy name for users to come up with directly, we need examples, quickstarts and user guides to guide them.

jason-dai replied:
It does not have to be `trainer.compile`; maybe something similar to `ray.distribtued`, such as `nano.pl_module`

zhentaocc commented:
> I don't think we should implementation multiple interfaces for the same use case - it just confuses the users; instead, just provide one single interface for one use case.

Sure, we need to decide which usage is much more common and friendly to users. And at the same time onnxruntime support should be aligned if we choose to wrap a torch model instance instead of a torch class.
But in the other way, I implemented another function which can either be a decorator or a usual function on a torch model instance:
```python
@to_lightning(loss, torch.optim.Adam, lr=0.01)
class Net(nn.Module):
    pass
pl_model = Net()
```
or
```python
class Net(nn.Module):
    pass
pl_model = to_lightning(loss, torch.optim.Adam, lr=0.01(Net())
```
I suppose we may have more extensions coming on pytorchlightning model. Maybe we can wrap all these extensions into one so user don't have to decide which function to use:
```python
def composed(*decs):
    def deco(f):
        for dec in reversed(decs):
            f = dec(f)
        return f
    return deco

def preprocess(loss=None, optimizer=None, config=None, onnx=True):
    return composed(
        onnxruntime(onnx),
        to_lightning(loss, optimizer, **config)
    )
pl_model = preprocess(loss, torch.optim.Adam, {"lr": 0.01}, onnx=True)(model)
```
What's your thoughts on this, providing a unified interface or just choose one from it?

jason-dai commented:
> > I don't think we should implementation multiple interfaces for the same use case - it just confuses the users; instead, just provide one single interface for one use case.
> 
> Sure, we need to decide which usage is much more common and friendly to users. And at the same time onnxruntime support should be aligned if we choose to wrap a torch model instance instead of a torch class. But in the other way, I implemented another function which can either be a decorator or a usual function on a torch model instance:
> 
> ```python
> @to_lightning(loss, torch.optim.Adam, lr=0.01)
> class Net(nn.Module):
>     pass
> pl_model = Net()
> ```
> 
> or
> 
> ```python
> class Net(nn.Module):
>     pass
> pl_model = to_lightning(loss, torch.optim.Adam, lr=0.01(Net())
> ```
> 
> I suppose we may have more extensions coming on pytorchlightning model. Maybe we can wrap all these extensions into one so user don't have to decide which function to use:
> 
> ```python
> def composed(*decs):
>     def deco(f):
>         for dec in reversed(decs):
>             f = dec(f)
>         return f
>     return deco
> 
> def preprocess(loss=None, optimizer=None, config=None, onnx=True):
>     return composed(
>         onnxruntime(onnx),
>         to_lightning(loss, optimizer, **config)
>     )
> pl_model = preprocess(loss, torch.optim.Adam, {"lr": 0.01}, onnx=True)(model)
> ```
> 
> We can also integrate this function into `trainer.compile()`. What's your thoughts on this, providing a unified interface or just choose one from it?

`pl_model = to_lightning(loss, torch.optim.Adam, lr=0.01(Net())` - is this correct? I think we want to have something similar to `ray.remote`, but support below cases:

* Class definition
  ```python
  Class Net(torch.nn.module):
      ...
  model = Net()
  ```

* Model instance
  ```python
  model = torch.vision.ResNet18()
  ```

* Model creator function
  ```python
  def model_creator():
     model = torch.vision.ResNet18()
     return model
  ```

Need a consistent API for all these use cases.

zhentaocc commented:
Having discussed with @TheaperDeng @yangw1234 , currently we have 3 solutions:

1.  `trainer.compile(model, loss, optimizer)` and return a lightning module
2.  `trainer.compile(model, loss, optimizer)` and return a lightning module. With extra flag to track if `trainer.compile` is called or not, we forced user to use this function before `fit()`. 
3.  Bind model to trainer, convert pytorch model to lightning when trainer is initiated and no need to pass model to `fit()`.

We prefer option 2 to make the conversion as a fixed process and giving proper warning and error to inform users of the correct usage. Finally user need to do:
`create model --> create loss,optim --> pl_model=compile(model, loss, optim) -> fit(pl_model, ...)`
And this can cover all 3 use cases mentioned above.

 @jason-dai What's your thoughts?

jason-dai commented:
> Having discussed with @TheaperDeng @yangw1234 , currently we have 3 solutions:
> 
> 1. `trainer.compile(model, loss, optimizer)` and return a lightning module
> 2. `trainer.compile(model, loss, optimizer)` and return a lightning module. With extra flag to track if `trainer.compile` is called or not, we forced user to use this function before `fit()`.
> 3. Bind model to trainer, convert pytorch model to lightning when trainer is initiated and no need to pass model to `fit()`.
> 
> We prefer option 2 to make the conversion as a fixed process and giving proper warning and error to inform users of the correct usage. Finally user need to do: `create model --> create loss,optim --> pl_model=compile(model, loss, optim) -> fit(pl_model, ...)` And this can cover all 3 use cases mentioned above.
> 
> The name of this `trainer.compile(model, loss, optimizer)` is to be decided so that no confusion to users.
> 
> @jason-dai What's your thoughts?

I think we only need `compile` for PyTorch model; otherwise the user needs to change his or her PL code.

zhentaocc commented:
> > Having discussed with @TheaperDeng @yangw1234 , currently we have 3 solutions:
> > 
> > 1. `trainer.compile(model, loss, optimizer)` and return a lightning module
> > 2. `trainer.compile(model, loss, optimizer)` and return a lightning module. With extra flag to track if `trainer.compile` is called or not, we forced user to use this function before `fit()`.
> > 3. Bind model to trainer, convert pytorch model to lightning when trainer is initiated and no need to pass model to `fit()`.
> > 
> > We prefer option 2 to make the conversion as a fixed process and giving proper warning and error to inform users of the correct usage. Finally user need to do: `create model --> create loss,optim --> pl_model=compile(model, loss, optim) -> fit(pl_model, ...)` And this can cover all 3 use cases mentioned above.
> > The name of this `trainer.compile(model, loss, optimizer)` is to be decided so that no confusion to users.
> > @jason-dai What's your thoughts?
> 
> I think we only need `compile` for PyTorch model; otherwise the user needs to change his or her PL code.

The main reason why we want to make `compile` a necessary step is that considering adding other extensions on Lightning (e.g. onnxruntime support, maybe following we'll have other extensions to add), integrating all these to `compile` relieves users from calling many interfaces to use different extensions. Even though user has already created a lightning model, `compile` can still do nothing and return the model. 
```python
pl_model = trainer.compile(torch.vision.ResNet18(),  loss, optimizer, ...)
pl_model = trainer.compile(LightningModule(...)) # This does nothing and returns LightningModule(...)
```
Above are all legal for users to use so they don't have to change their lightning code. By this way, they can be warned to use `complie` if they need any extra extension in nano as well as converting to a pl model.

jason-dai commented:
> > > Having discussed with @TheaperDeng @yangw1234 , currently we have 3 solutions:
> > > 
> > > 1. `trainer.compile(model, loss, optimizer)` and return a lightning module
> > > 2. `trainer.compile(model, loss, optimizer)` and return a lightning module. With extra flag to track if `trainer.compile` is called or not, we forced user to use this function before `fit()`.
> > > 3. Bind model to trainer, convert pytorch model to lightning when trainer is initiated and no need to pass model to `fit()`.
> > > 
> > > We prefer option 2 to make the conversion as a fixed process and giving proper warning and error to inform users of the correct usage. Finally user need to do: `create model --> create loss,optim --> pl_model=compile(model, loss, optim) -> fit(pl_model, ...)` And this can cover all 3 use cases mentioned above.
> > > The name of this `trainer.compile(model, loss, optimizer)` is to be decided so that no confusion to users.
> > > @jason-dai What's your thoughts?
> > 
> > 
> > I think we only need `compile` for PyTorch model; otherwise the user needs to change his or her PL code.
> 
> The main reason why we want to make `compile` a necessary step is that considering adding other extensions on Lightning (e.g. onnxruntime support, maybe following we'll have other extensions to add), integrating all these to `compile` relieves users from calling many interfaces to use different extensions. Even though user has already created a lightning model, `compile` can still do nothing and return the model.
> 
> ```python
> pl_model = trainer.compile(torch.vision.ResNet18(),  loss, optimizer, ...)
> pl_model = trainer.compile(LightningModule(...)) # This does nothing and returns LightningModule(...)
> ```
> 
> Above are all legal for users to use so they don't have to change their lightning code. By this way, they can be warned to use `complie` if they need any extra extension in nano as well as converting to a pl model.

It's OK to use `compile` to add nano-specific extensions; however, the user should not be required to call `compile` is he or she just has standard PL code and wants to use `nano` for transparent acceleration.

zhentaocc commented:
As discussed above, make modifications:

1. Refactored `to_lightning` to class `LightningModuleFromTorch`
2. Added unit test to test `LightningModuleFromTorch`
3. Added static method `Trainer.compile(...)` since `self` is not used
4. Added unit test to test `Trainer.compile(...)`

TheaperDeng commented on python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py:
```diff
@@ -104,3 +108,23 @@ def __init__(self, num_processes: int = 1,
+    def compile(model: nn.Module, loss: _Loss = None, optimizer: torch.optim = None):
+        """
+        Compile a pytorch model into a pytorch-lightning model and return it.
+        Args:
```
maybe write this in the format as https://github.com/intel-analytics/BigDL/blob/branch-2.0/python/chronos/src/bigdl/chronos/autots/tspipeline.py#L47-L60

zhentaocc replied:
Applied your suggestion. Thanks.

TheaperDeng commented on python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py:
```diff
@@ -104,3 +108,23 @@ def __init__(self, num_processes: int = 1,
             super().__init__(accelerator=accelerator,
                              plugins=[plugin], *args, **kwargs)
+
+    @staticmethod
```
I think we can make it a member function

zhentaocc replied:
Static method is more reasonable, I think, after all we are not using any Trainer instance in compile. And we can do either:
1.  `trainer.compile(model, loss, optimizer)` 
2.  `Trainer.compile(model, loss, optimizer)`

TheaperDeng commented on python/nano/src/bigdl/nano/pytorch/lightning.py:
```diff
@@ -0,0 +1,70 @@
+        self.loss = loss
+        self.optimizer = optimizer
+
+    def copy(self, torch_model):
```
make this an internal method `_copy`?

zhentaocc replied:
Cool.

jason-dai commented on python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py:
```diff
@@ -104,3 +108,24 @@ def __init__(self, num_processes: int = 1,
+    @staticmethod
+    def compile(model: nn.Module, loss: _Loss = None, optimizer: torch.optim = None):
+        """
+        Compile a pytorch model into a pytorch-lightning model and return it.
```
the comment is incorrect

zhentaocc replied:
Please review modified docstring. Any further suggestions to properly describe this function?

jason-dai commented on python/nano/src/bigdl/nano/pytorch/trainer/Trainer.py:
```diff
@@ -104,3 +108,24 @@ def __init__(self, num_processes: int = 1,
+                model.loss = loss
+            if optimizer:
+                model.optimizer = optimizer
+            return model
```
Why do we want to support this? Can we guarantee this is correct if `configure_optimizers` returns a different optimizer?

zhentaocc replied:
Correct. Added assertion to check if the inputs are correct.

zhentaocc commented:
Passed tests. It's ready now.

yangw1234 commented on python/nano/src/bigdl/nano/pytorch/lightning.py:
```diff
@@ -0,0 +1,66 @@
+        self.loss = loss
+        self.optimizer = optimizer
+
+    def _copy(self, torch_model: nn.Module):
```
Is it safe to do this? 
Looks a little hacking to me. 
Should we stick to the public API pytorch-lightning provides?

jason-dai replied:
Why not make `torch_model` a member variable?

zhentaocc replied:
We have `load_state_dict` in `nn.Module`. So I'm concerned if we make it as a member variable user won't be able to load a pretrained torch model to pytorch-lightning. We can still assume that user has already loaded the parameters before creating this pl model. But I'm not sure if all other functions work fine.

zhentaocc replied:
Having checked Pytorch API doc, it seems only `load_state_dict` is influenced since the module dict will be changed if make it a member variable. I think we have 3 options:

1.  Make pytorch model as a member variable and assume user has already loaded the model.
2.  Make pytorch model as a member variable and override `load_state_dict`.
3.  Move `_copy` into `__init__`. (Not sure if this seems not that hacking.)

jason-dai replied:
We can do #2.

Another option: can we use `__getattr__`, similar to https://github.com/intel-analytics/BigDL/blob/branch-2.0/python/friesian/src/bigdl/friesian/feature/table.py#L663?

zhentaocc replied:
No idea how `getattr` works in this case. Current implementation will try to load self.model firstly and then try to load self if failed.

```python
try:
    self.model.load_state_dict(state_dict)
except RuntimeError:
    super().load_state_dict(state_dict)
```
And I added two unit tests to make sure they work in both cases.

jason-dai replied:
> No idea how `getattr` works in this case. Current implementation will try to load self.model firstly and then try to load self if failed.
> 
> ```python
> try:
>     self.model.load_state_dict(state_dict)
> except RuntimeError:
>     super().load_state_dict(state_dict)
> ```
> 
> And I added two unit tests to make sure they work in both cases.

Exactly what `Table` does:
``` 
   def __getattr__(self, name):
        return self.model.__getattr__(name)
```

zhentaocc replied:
I understand now. If we use this method, it works when I load state_dict from a torch model checkpoint. 
Let's say keys are "conv.weight", "conv.bias", and keys for lightning module are "model.conv.weight", "model.conv.bias", it works.
Otherwise if I stored a pytorch-lightning model, its keys are actually "model.conv.weight", "model.conv.bias", then it failed. 
If we just want to resolve `load_state_dict`, possibly:

1.  We accept that user can load state dict from either a torch model or pl model.
Simplest way it to try loading into `self.model` and then `self`, otherwise raise error.

2.  We assume user always load state dict from torch model before calling `compile` or `LightningModuleFromTorch`.
Then we don't have to do anything.

zhentaocc replied:
> I understand now. If we use this method, it works when I load state_dict from a torch model checkpoint. Let's say keys are `conv.weight`, `conv.bias`, and keys for lightning module are `model.conv.weight`, `model.conv.bias`, it works. Otherwise if I stored a pytorch-lightning model, its keys are actually `model.conv.weight`, `model.conv.bias`, then it failed. If we just want to resolve `load_state_dict`, possibly:
> 
> 1. We accept that user can load state dict from either a torch model or pl model.
>    Simplest way it to try loading into `self.model` and then `self`, otherwise raise error.
> 2. We assume user always load state dict from torch model before calling `compile` or `LightningModuleFromTorch`.
>    Then we don't have to do anything.

Main difference is keys for a pl model have 'model.' prefix.

jason-dai replied:
> I understand now. If we use this method, it works when I load state_dict from a torch model checkpoint. Let's say keys are `conv.weight`, `conv.bias`, and keys for lightning module are `model.conv.weight`, `model.conv.bias`, it works. Otherwise if I stored a pytorch-lightning model, its keys are actually `model.conv.weight`, `model.conv.bias`, then it failed. If we just want to resolve `load_state_dict`, possibly:
> 
> 1. We accept that user can load state dict from either a torch model or pl model.
>    Simplest way it to try loading into `self.model` and then `self`, otherwise raise error.
> 2. We assume user always load state dict from torch model before calling `compile` or `LightningModuleFromTorch`.
>    Then we don't have to do anything.

I think the point is:
1) We can override whatever methods (such as `load_state_dict`, `forward`, etc.) that we would like to in `LightningModuleFromTorch`
2) For methods defined in the original python module that we do not override, we may use `__getattr__` so that these methods can still work (if needed at all)

zhentaocc replied:
Got it. We discussed and preferred 2nd option.
Everything looks good now.

yangw1234 commented on python/nano/src/bigdl/nano/pytorch/lightning.py:
```diff
@@ -0,0 +1,66 @@
+            setattr(self, name, child)
+        setattr(self, "forward", torch_model.forward)
+
+    def _forward(self, batch):
```
could you document why you override this method?

yangw1234 commented on python/nano/src/bigdl/nano/pytorch/lightning.py:
```diff
@@ -0,0 +1,66 @@
+
+    def training_step(self, batch, batch_idx):
+        x, y = batch
+        y_hat = self._forward(batch)
```
I am sure I understand this method. 
Could we just call `self.forward` in `self._forward` and pass `x` to it?

zhentaocc replied:
We have a few forward  with multiple inputs instead of one. Typically in Seq2Seq, during training the input can be (x, y) to use y as a prior, but input (x) during inference. Another case is that input is (x1,x2,x3,...,y). So in `self._forward` I'm trying to resolve this issue by getting the number of arguments. Another way is call `self.forward(batch)` and we have to handle some extra input like y when defining forward function, for example, using `forward(x, *args)`.

yangw1234 replied:
for teacher forcing, can we let `Seq2Seq` override `train_step`. @TheaperDeng

yangw1234 commented:
LGTM

zhentaocc commented:
Jenkins passed.
http://10.112.231.51:18888/job/ZOO-PR-NanoTests/68/

yangw1234 commented:
http://10.112.231.51:18888/job/ZOO-PR-NanoTests/74/
